File: Pinball App 2/Pinball App 2/practice/PracticePRPAProfileRemoteSupport.swift
import Foundation

enum PRPAPublicProfileService {
    static func fetchProfile(playerID: String) async throws -> PRPAPlayerProfile {
        guard var components = URLComponents(string: "https://punkrockpinball.com/player/") else {
            throw PRPAProfileError.invalidURL
        }
        components.queryItems = [URLQueryItem(name: "prp_id", value: playerID)]
        guard let url = components.url else {
            throw PRPAProfileError.invalidURL
        }

        let (data, response) = try await URLSession.shared.data(from: url)
        guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode) else {
            throw PRPAProfileError.networkFailure
        }
        guard let html = String(data: data, encoding: .utf8) else {
            throw PRPAProfileError.unreadableResponse
        }

        return try parseProfile(html: html, playerID: playerID)
    }

    private static func parseProfile(html: String, playerID: String) throws -> PRPAPlayerProfile {
        guard html.localizedCaseInsensitiveContains("prpp-player-results") else {
            throw PRPAProfileError.parseFailure
        }

        let displayName = firstMatch(in: html, pattern: #"<h1[^>]*>\s*(.*?)\s*</h1>"#)?
            .prpaCleanedHTMLText() ?? "PRPA Player"

        let scenesSection = html.prpaSlice(from: #"<div class="prpp-summary-scenes">"#, to: #"</div>"#) ?? ""
        let scenes = matches(
            in: scenesSection,
            pattern: #"<a [^>]*>([^<]+)</a>\s*<strong[^>]*>([^<]+)</strong>"#
        )
        .compactMap { groups -> PRPASceneStanding? in
            guard groups.count >= 2 else { return nil }
            return PRPASceneStanding(
                name: groups[0].prpaCleanedHTMLText(),
                rank: groups[1].prpaCleanedHTMLText()
            )
        }

        let tableSection = html.prpaSlice(from: #"<table class="prpp-table widefat striped">"#, to: #"</table>"#) ?? ""
        let tournaments = matches(
            in: tableSection,
            pattern: #"<tr>\s*<td>(.*?)</td>\s*<td>([^<]+)</td>\s*<td>([^<]+)</td>\s*<td>([^<]+)</td>\s*</tr>"#
        )
        .compactMap { groups -> PRPARecentTournament? in
            guard groups.count >= 4 else { return nil }
            let tournamentCell = groups[0]
            let dateLabel = groups[1].prpaCleanedHTMLText()
            guard let date = resultDateFormatter.date(from: dateLabel) else { return nil }
            let name = firstMatch(in: tournamentCell, pattern: #"<a [^>]*>([^<]+)</a>"#)?
                .prpaCleanedHTMLText() ?? tournamentCell.prpaCleanedHTMLText()
            let eventType = firstMatch(in: tournamentCell, pattern: #"<span class="prpp-badge[^"]*">([^<]+)</span>"#)?
                .prpaCleanedHTMLText()
            return PRPARecentTournament(
                name: name,
                eventType: eventType,
                date: date,
                dateLabel: dateLabel,
                placement: groups[2].prpaCleanedHTMLText(),
                pointsGained: groups[3].prpaCleanedHTMLText()
            )
        }
        .sorted { lhs, rhs in
            if lhs.date == rhs.date {
                return lhs.name < rhs.name
            }
            return lhs.date > rhs.date
        }

        return PRPAPlayerProfile(
            playerID: playerID,
            displayName: displayName,
            openPoints: summaryValue(in: html, label: "Open PRPA Points:") ?? "-",
            eventsPlayed: summaryValue(in: html, label: "Events Played:") ?? "-",
            openRanking: summaryValue(in: html, label: "Open Ranking:") ?? "-",
            averagePointsPerEvent: summaryValue(in: html, label: "Average Points/Event:") ?? "-",
            bestFinish: summaryValue(in: html, label: "Best Finish (by points):") ?? "-",
            worstFinish: summaryValue(in: html, label: "Worst Finish (by points):") ?? "-",
            ifpaPlayerID: summaryValue(in: html, label: "IFPA ID:"),
            lastEventDate: tournaments.first?.dateLabel,
            scenes: scenes,
            recentTournaments: Array(tournaments.prefix(3))
        )
    }

    private static let resultDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "M/d/yyyy - h:mma"
        formatter.amSymbol = "am"
        formatter.pmSymbol = "pm"
        return formatter
    }()

    private static func summaryValue(in html: String, label: String) -> String? {
        let escapedLabel = NSRegularExpression.escapedPattern(for: label)
        let pattern = "<li[^>]*>\\s*<strong>\\s*\(escapedLabel)\\s*</strong>\\s*(?:<span[^>]*>)?\\s*([^<]+)"
        return firstMatch(in: html, pattern: pattern)?.prpaCleanedHTMLText()
    }

    private static func firstMatch(in text: String, pattern: String, group: Int = 1) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive, .dotMatchesLineSeparators]) else {
            return nil
        }
        let nsRange = NSRange(text.startIndex..., in: text)
        guard let match = regex.firstMatch(in: text, options: [], range: nsRange),
              let range = Range(match.range(at: group), in: text) else {
            return nil
        }
        return String(text[range])
    }

    private static func matches(in text: String, pattern: String) -> [[String]] {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive, .dotMatchesLineSeparators]) else {
            return []
        }
        let nsRange = NSRange(text.startIndex..., in: text)
        return regex.matches(in: text, options: [], range: nsRange).map { match in
            (1..<match.numberOfRanges).compactMap { index in
                guard let range = Range(match.range(at: index), in: text) else { return nil }
                return String(text[range])
            }
        }
    }
}

enum PRPAProfileError: LocalizedError {
    case invalidURL
    case networkFailure
    case unreadableResponse
    case parseFailure

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "The PRPA profile URL could not be created."
        case .networkFailure:
            return "The PRPA profile could not be loaded right now."
        case .unreadableResponse:
            return "The PRPA profile response could not be read."
        case .parseFailure:
            return "The public PRPA profile layout did not match the expected format."
        }
    }
}

private extension String {
    func prpaCleanedHTMLText() -> String {
        self
            .replacingOccurrences(of: "<[^>]+>", with: " ", options: .regularExpression)
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&#8211;", with: "-")
            .replacingOccurrences(of: "&ndash;", with: "-")
            .replacingOccurrences(of: "&#8217;", with: "'")
            .replacingOccurrences(of: "&#039;", with: "'")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#8220;", with: "\"")
            .replacingOccurrences(of: "&#8221;", with: "\"")
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func prpaSlice(from startMarker: String, to endMarker: String) -> String? {
        guard let startRange = range(of: startMarker) else { return nil }
        guard let endRange = range(of: endMarker, range: startRange.upperBound..<endIndex) else { return nil }
        return String(self[startRange.upperBound..<endRange.lowerBound])
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/practice/PracticePrpaProfileRemoteSupport.kt
package com.pillyliu.pinprofandroid.practice

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URL
import java.time.LocalDateTime
import java.time.format.DateTimeFormatterBuilder
import java.util.Locale

internal object PrpaPublicProfileService {
    private val resultDateFormatter = DateTimeFormatterBuilder()
        .parseCaseInsensitive()
        .appendPattern("M/d/yyyy - h:mma")
        .toFormatter(Locale.US)

    suspend fun fetchProfile(playerID: String): PrpaPlayerProfile = withContext(Dispatchers.IO) {
        val html = URL("https://punkrockpinball.com/player/?prp_id=$playerID").readText()
        parseProfile(playerID, html)
    }

    private fun parseProfile(playerID: String, html: String): PrpaPlayerProfile {
        if (!html.contains("prpp-player-results", ignoreCase = true)) {
            throw IllegalStateException("The public PRPA profile layout did not match the expected format.")
        }

        val displayName = firstMatch(html, """<h1[^>]*>\s*(.*?)\s*</h1>""")?.cleanedPrpaHtmlText() ?: "PRPA Player"
        val scenesSection = html.slicePrpa(
            from = """<div class="prpp-summary-scenes">""",
            to = """</div>""",
        ).orEmpty()
        val scenes = allMatches(scenesSection, """<a [^>]*>([^<]+)</a>\s*<strong[^>]*>([^<]+)</strong>""")
            .mapNotNull { groups ->
                if (groups.size < 2) return@mapNotNull null
                PrpaSceneStanding(
                    name = groups[0].cleanedPrpaHtmlText(),
                    rank = groups[1].cleanedPrpaHtmlText(),
                )
            }

        val tableSection = html.slicePrpa(
            from = """<table class="prpp-table widefat striped">""",
            to = """</table>""",
        ).orEmpty()
        val tournaments = allMatches(
            tableSection,
            """<tr>\s*<td>(.*?)</td>\s*<td>([^<]+)</td>\s*<td>([^<]+)</td>\s*<td>([^<]+)</td>\s*</tr>""",
        ).mapNotNull { groups ->
            if (groups.size < 4) return@mapNotNull null
            val tournamentCell = groups[0]
            val dateLabel = groups[1].cleanedPrpaHtmlText()
            val date = runCatching { LocalDateTime.parse(dateLabel, resultDateFormatter) }.getOrNull() ?: return@mapNotNull null
            val name = firstMatch(tournamentCell, """<a [^>]*>([^<]+)</a>""")?.cleanedPrpaHtmlText()
                ?: tournamentCell.cleanedPrpaHtmlText()
            val eventType = firstMatch(tournamentCell, """<span class="prpp-badge[^"]*">([^<]+)</span>""")?.cleanedPrpaHtmlText()
            PrpaRecentTournament(
                name = name,
                eventType = eventType,
                date = date,
                dateLabel = dateLabel,
                placement = groups[2].cleanedPrpaHtmlText(),
                pointsGained = groups[3].cleanedPrpaHtmlText(),
            )
        }.sortedByDescending { it.date }

        return PrpaPlayerProfile(
            playerID = playerID,
            displayName = displayName,
            openPoints = summaryValue(html, "Open PRPA Points:") ?: "-",
            eventsPlayed = summaryValue(html, "Events Played:") ?: "-",
            openRanking = summaryValue(html, "Open Ranking:") ?: "-",
            averagePointsPerEvent = summaryValue(html, "Average Points/Event:") ?: "-",
            bestFinish = summaryValue(html, "Best Finish (by points):") ?: "-",
            worstFinish = summaryValue(html, "Worst Finish (by points):") ?: "-",
            ifpaPlayerID = summaryValue(html, "IFPA ID:"),
            lastEventDate = tournaments.firstOrNull()?.dateLabel,
            scenes = scenes,
            recentTournaments = tournaments.take(3),
        )
    }

    private fun summaryValue(text: String, label: String): String? {
        val pattern = """<li[^>]*>\s*<strong>\s*${Regex.escape(label)}\s*</strong>\s*(?:<span[^>]*>)?\s*([^<]+)"""
        return firstMatch(text, pattern)?.cleanedPrpaHtmlText()
    }

    private fun firstMatch(text: String, pattern: String, group: Int = 1): String? {
        return Regex(pattern, setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))
            .find(text)
            ?.groupValues
            ?.getOrNull(group)
    }

    private fun allMatches(text: String, pattern: String): List<List<String>> {
        return Regex(pattern, setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))
            .findAll(text)
            .map { it.groupValues.drop(1) }
            .toList()
    }
}

private fun String.cleanedPrpaHtmlText(): String {
    return this
        .replace(Regex("<[^>]+>"), " ")
        .replace("&amp;", "&")
        .replace("&nbsp;", " ")
        .replace("&#8211;", "-")
        .replace("&ndash;", "-")
        .replace("&#8217;", "'")
        .replace("&#039;", "'")
        .replace("&quot;", "\"")
        .replace(Regex("\\s+"), " ")
        .trim()
}

private fun String.slicePrpa(from: String, to: String): String? {
    val startIndex = indexOf(from)
    if (startIndex < 0) return null
    val contentStart = startIndex + from.length
    val endIndex = indexOf(to, startIndex = contentStart)
    if (endIndex < 0) return null
    return substring(contentStart, endIndex)
}
